jupyter_ai_persona_manager/handlers.py

```diff
@@ -1,14 +1,82 @@
 import json
+import mimetypes
+import os
+from functools import lru_cache
+from typing import Optional
 
-from jupyter_server.base.handlers import APIHandler
+from jupyter_server.base.handlers import JupyterHandler
 import tornado
 
-class RouteHandler(APIHandler):
+
+class RouteHandler(JupyterHandler):
     # The following decorator should be present on all verb methods (head, get, post,
     # patch, put, delete, options) to ensure only authorized user can request the
     # Jupyter server
     @tornado.web.authenticated
     def get(self):
+        self.set_header("Content-Type", "application/json")
         self.finish(json.dumps({
             "data": "This is /jupyter-ai-persona-manager/get-example endpoint!"
         }))
+
+
+class AvatarHandler(JupyterHandler):
+    """
+    Handler for serving persona avatar files.
+
+    Looks up avatar files through the PersonaManager to find the correct file path,
+    then serves the file with appropriate content-type headers.
+    """
+
+    @tornado.web.authenticated
+    async def get(self, filename: str):
+        """Serve an avatar file by filename."""
+        # Get the avatar file path from persona managers
+        avatar_path = self._find_avatar_file(filename)
+
+        if avatar_path is None:
+            raise tornado.web.HTTPError(404, f"Avatar file not found: {filename}")
+
+        # Serve the file
+        try:
+            # Set content type based on file extension
+            content_type, _ = mimetypes.guess_type(avatar_path)
+            if content_type:
+                self.set_header("Content-Type", content_type)
+
+            # Read and serve the file
+            with open(avatar_path, 'rb') as f:
+                content = f.read()
+                self.write(content)
+
+            await self.finish()
+        except Exception as e:
+            self.log.error(f"Error serving avatar file {filename}: {e}")
+            raise tornado.web.HTTPError(500, f"Error serving avatar file: {str(e)}")
+
+    @lru_cache(maxsize=128)
```

Comment:
This cache will be keyed on both `self` and `filename`. Because this is on the handler you will get many of these caches rather than a small number that will do the caching you want. Can you extract into a function that does the caching you want?

Reply:
Excellent catch\! You're absolutely right - the `lru_cache` on an instance method would create a new cache for each request handler instance, making it useless.

I've implemented a module-level cache instead:
- `_avatar_cache: Dict[str, str]` dictionary at module level
- `build_avatar_cache()` function called when personas are initialized or refreshed
- `_find_avatar_file()` now does O(1) dictionary lookup instead of iterating
- Cache is automatically rebuilt when `/refresh-personas` is called

This addresses your performance concern about walking all personas on each request.